libsql-storage-server/src/fdb_store.rs

```diff
@@ -40,77 +75,82 @@ impl FDBFrameStore {
         frame_no: u64,
         frame: Frame,
     ) {
-        let frame_data_key = format!("{}/f/{}/f", namespace, frame_no);
-        let frame_page_key = format!("{}/f/{}/p", namespace, frame_no);
-        let page_key = format!("{}/p/{}", namespace, frame.page_no);
-
-        txn.set(&frame_data_key.as_bytes(), &frame.data);
-        txn.set(&frame_page_key.as_bytes(), &pack(&frame.page_no));
-        txn.set(&page_key.as_bytes(), &pack(&frame_no));
+        let frame_data_key = frame_key(namespace, frame_no);
+        let frame_page_key = frame_page_key(namespace, frame_no);
+        let page_key = page_key(namespace, frame.page_no);
+        let page_frame_idx = page_index_key(namespace, frame.page_no, frame_no);
+
+        txn.set(&frame_data_key, &frame.data);
+        txn.set(&frame_page_key, &pack(&frame.page_no));
+        txn.set(&page_key, &pack(&frame_no));
+        txn.set(&page_frame_idx, &pack(&""));
     }
 }
 
 #[async_trait]
 impl FrameStore for FDBFrameStore {
-    async fn insert_frames(&self, namespace: &str, _max_frame_no: u64, frames: Vec<Frame>) -> u64 {
-        let max_frame_key = format!("{}/max_frame_no", namespace);
+    async fn insert_frames(
+        &self,
+        namespace: &str,
+        max_frame_no: u64,
+        frames: Vec<Frame>,
+    ) -> Result<u64, Error> {
         let db = foundationdb::Database::default().unwrap();
         let txn = db.create_trx().expect("unable to create transaction");
         let mut frame_no = self.get_max_frame_no(&txn, namespace).await;
-
+        if frame_no != max_frame_no {
+            return Err(WriteConflict);
+        }
```

Comment:
how does fdb actually handles conflict? we could very well have to writer submitting a transaction with the same max_frame_no at the same time, no?

Reply:
Yes, but only one of them will succeed. 

FDB provides SSI - Strict Serialisable Isolation. If a transaction reads/writes some value, then another transaction cannot change it in between.